File: Java/src/main/java/com/ilinkai/weixin/messaging/MessageSendService.java
package com.ilinkai.weixin.messaging;

import com.ilinkai.weixin.WeixinApiClient;
import com.ilinkai.weixin.WeixinApiException;
import com.ilinkai.weixin.cdn.CdnException;
import com.ilinkai.weixin.media.MediaUploadService;
import com.ilinkai.weixin.models.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.nio.file.Paths;
import java.util.*;

/** 消息发送服务 */
public class MessageSendService {
    private static final Logger logger = LoggerFactory.getLogger(MessageSendService.class);
    private final WeixinApiClient apiClient;
    private final MediaUploadService uploadService;

    public MessageSendService(WeixinApiClient apiClient, String cdnBaseUrl) {
        this.apiClient = apiClient;
        this.uploadService = new MediaUploadService(apiClient, cdnBaseUrl);
    }

    private static String generateClientId() {
        return "ilinkai-weixin-" + UUID.randomUUID().toString().replace("-", "");
    }

    public String sendText(String toUserId, String text, String contextToken)
            throws WeixinApiException, IOException {
        if (contextToken == null || contextToken.isEmpty())
            throw new IllegalArgumentException("contextToken is required");

        String clientId = generateClientId();
        List<MessageItem> items = new ArrayList<>();
        if (text != null && !text.isEmpty()) {
            MessageItem item = new MessageItem();
            item.setType(Enums.MessageItemType.TEXT);
            MessageItems.TextItem ti = new MessageItems.TextItem();
            ti.setText(text);
            item.setTextItem(ti);
            items.add(item);
        }

        WeixinMessage msg = new WeixinMessage();
        msg.setFromUserId("");
        msg.setToUserId(toUserId);
        msg.setClientId(clientId);
        msg.setMessageType(Enums.MessageType.BOT);
        msg.setMessageState(Enums.MessageState.FINISH);
        msg.setItemList(items);
        msg.setContextToken(contextToken);

        ApiModels.SendMessageRequest request = new ApiModels.SendMessageRequest();
        request.setMessage(msg);
        apiClient.sendMessage(request);
        return clientId;
    }

    public String sendImage(String toUserId, String filePath, String contextToken, String caption)
            throws WeixinApiException, IOException, CdnException {
        if (contextToken == null || contextToken.isEmpty())
            throw new IllegalArgumentException("contextToken is required");

        MediaUploadService.UploadedFileInfo uploaded = uploadService.uploadImage(filePath, toUserId);
        MessageItem imageItem = new MessageItem();
        imageItem.setType(Enums.MessageItemType.IMAGE);
        MessageItems.ImageItem img = new MessageItems.ImageItem();
        CdnMedia media = new CdnMedia();
        media.setEncryptQueryParam(uploaded.downloadEncryptedQueryParam);
        media.setAesKey(Base64.getEncoder().encodeToString(hexStringToBytes(uploaded.aesKey)));
        media.setEncryptType(1);
        img.setMedia(media);
        img.setMidSize(uploaded.fileSizeCiphertext);
        imageItem.setImageItem(img);
        return sendMediaItems(toUserId, caption, imageItem, contextToken, "SendImage");
    }

    public String sendVideo(String toUserId, String filePath, String contextToken, String caption)
            throws WeixinApiException, IOException, CdnException {
        if (contextToken == null || contextToken.isEmpty())
            throw new IllegalArgumentException("contextToken is required");

        MediaUploadService.UploadedFileInfo uploaded = uploadService.uploadVideo(filePath, toUserId);
        MessageItem videoItem = new MessageItem();
        videoItem.setType(Enums.MessageItemType.VIDEO);
        MessageItems.VideoItem vi = new MessageItems.VideoItem();
        CdnMedia media = new CdnMedia();
        media.setEncryptQueryParam(uploaded.downloadEncryptedQueryParam);
        media.setAesKey(Base64.getEncoder().encodeToString(hexStringToBytes(uploaded.aesKey)));
        media.setEncryptType(1);
        vi.setMedia(media);
        vi.setVideoSize(uploaded.fileSizeCiphertext);
        videoItem.setVideoItem(vi);
        return sendMediaItems(toUserId, caption, videoItem, contextToken, "SendVideo");
    }

    public String sendFile(String toUserId, String filePath, String contextToken, String caption)
            throws WeixinApiException, IOException, CdnException {
        if (contextToken == null || contextToken.isEmpty())
            throw new IllegalArgumentException("contextToken is required");

        String fileName = Paths.get(filePath).getFileName().toString();
        MediaUploadService.UploadedFileInfo uploaded = uploadService.uploadFileAttachment(filePath, toUserId);
        MessageItem fileItem = new MessageItem();
        fileItem.setType(Enums.MessageItemType.FILE);
        MessageItems.FileItem fi = new MessageItems.FileItem();
        CdnMedia media = new CdnMedia();
        media.setEncryptQueryParam(uploaded.downloadEncryptedQueryParam);
        media.setAesKey(Base64.getEncoder().encodeToString(hexStringToBytes(uploaded.aesKey)));
        media.setEncryptType(1);
        fi.setMedia(media);
        fi.setFileName(fileName);
        fi.setLength(String.valueOf(uploaded.fileSize));
        fileItem.setFileItem(fi);
        return sendMediaItems(toUserId, caption, fileItem, contextToken, "SendFile");
    }

    private String sendMediaItems(String toUserId, String text, MessageItem mediaItem,
                                   String contextToken, String label) throws WeixinApiException, IOException {
        List<MessageItem> items = new ArrayList<>();
        if (text != null && !text.isEmpty()) {
            MessageItem textItem = new MessageItem();
            textItem.setType(Enums.MessageItemType.TEXT);
            MessageItems.TextItem ti = new MessageItems.TextItem();
            ti.setText(text);
            textItem.setTextItem(ti);
            items.add(textItem);
        }
        items.add(mediaItem);

        String lastClientId = "";
        for (MessageItem item : items) {
            lastClientId = generateClientId();
            WeixinMessage msg = new WeixinMessage();
            msg.setFromUserId("");
            msg.setToUserId(toUserId);
            msg.setClientId(lastClientId);
            msg.setMessageType(Enums.MessageType.BOT);
            msg.setMessageState(Enums.MessageState.FINISH);
            msg.setItemList(Collections.singletonList(item));
            msg.setContextToken(contextToken);

            ApiModels.SendMessageRequest request = new ApiModels.SendMessageRequest();
            request.setMessage(msg);
            apiClient.sendMessage(request);
        }
        return lastClientId;
    }

    public void sendTyping(String ilinkUserId, String typingTicket, boolean isTyping)
            throws WeixinApiException, IOException {
        ApiModels.SendTypingRequest request = new ApiModels.SendTypingRequest();
        request.setILinkUserId(ilinkUserId);
        request.setTypingTicket(typingTicket);
        request.setStatus(isTyping ? Enums.TypingStatus.TYPING : Enums.TypingStatus.CANCEL);
        apiClient.sendTyping(request);
    }

    private static byte[] hexStringToBytes(String hex) {
        byte[] result = new byte[hex.length() / 2];
        for (int i = 0; i < result.length; i++) {
            result[i] = (byte) Integer.parseInt(hex.substring(i * 2, i * 2 + 2), 16);
        }
        return result;
    }
}

File: Python/ilinkai_weixin/messaging/message_send_service.py
"""消息发送服务"""
import base64
import logging
import os
import uuid
from typing import Optional

from ..models import (
    SendMessageRequest, WeixinMessage, MessageItem, MessageItemType,
    MessageType, MessageState, TypingStatus, SendTypingRequest,
    CdnMedia, TextItem, ImageItem, VideoItem, FileItem,
)
from ..media.media_upload_service import MediaUploadService

logger = logging.getLogger(__name__)


class MessageSendService:
    def __init__(self, api_client, cdn_base_url: str):
        self._api_client = api_client
        self._upload_service = MediaUploadService(api_client, cdn_base_url)

    @staticmethod
    def _generate_client_id() -> str:
        return f"ilinkai-weixin-{uuid.uuid4().hex}"

    def send_text(self, to_user_id: str, text: str, context_token: str) -> str:
        if not context_token:
            raise ValueError("contextToken is required")
        client_id = self._generate_client_id()
        items = []
        if text:
            ti = TextItem(text=text)
            items.append(MessageItem(type=MessageItemType.TEXT, text_item=ti))

        msg = WeixinMessage(
            from_user_id="", to_user_id=to_user_id, client_id=client_id,
            message_type=MessageType.BOT, message_state=MessageState.FINISH,
            item_list=items, context_token=context_token,
        )
        request = SendMessageRequest(message=msg)
        self._api_client.send_message(request)
        return client_id

    def send_image(self, to_user_id: str, file_path: str, context_token: str,
                   caption: Optional[str] = None) -> str:
        if not context_token:
            raise ValueError("contextToken is required")
        uploaded = self._upload_service.upload_image(file_path, to_user_id)
        media = CdnMedia(
            encrypt_query_param=uploaded.download_encrypted_query_param,
            aes_key=base64.b64encode(bytes.fromhex(uploaded.aes_key)).decode(),
            encrypt_type=1,
        )
        image_item = MessageItem(
            type=MessageItemType.IMAGE,
            image_item=ImageItem(media=media, mid_size=uploaded.file_size_ciphertext),
        )
        return self._send_media_items(to_user_id, caption, image_item, context_token, "SendImage")

    def send_video(self, to_user_id: str, file_path: str, context_token: str,
                   caption: Optional[str] = None) -> str:
        if not context_token:
            raise ValueError("contextToken is required")
        uploaded = self._upload_service.upload_video(file_path, to_user_id)
        media = CdnMedia(
            encrypt_query_param=uploaded.download_encrypted_query_param,
            aes_key=base64.b64encode(bytes.fromhex(uploaded.aes_key)).decode(),
            encrypt_type=1,
        )
        video_item = MessageItem(
            type=MessageItemType.VIDEO,
            video_item=VideoItem(media=media, video_size=uploaded.file_size_ciphertext),
        )
        return self._send_media_items(to_user_id, caption, video_item, context_token, "SendVideo")

    def send_file(self, to_user_id: str, file_path: str, context_token: str,
                  caption: Optional[str] = None) -> str:
        if not context_token:
            raise ValueError("contextToken is required")
        file_name = os.path.basename(file_path)
        uploaded = self._upload_service.upload_file_attachment(file_path, to_user_id)
        media = CdnMedia(
            encrypt_query_param=uploaded.download_encrypted_query_param,
            aes_key=base64.b64encode(bytes.fromhex(uploaded.aes_key)).decode(),
            encrypt_type=1,
        )
        file_item = MessageItem(
            type=MessageItemType.FILE,
            file_item=FileItem(media=media, file_name=file_name, length=str(uploaded.file_size)),
        )
        return self._send_media_items(to_user_id, caption, file_item, context_token, "SendFile")

    def _send_media_items(self, to_user_id: str, text: Optional[str],
                           media_item: MessageItem, context_token: str, label: str) -> str:
        items = []
        if text:
            ti = TextItem(text=text)
            items.append(MessageItem(type=MessageItemType.TEXT, text_item=ti))
        items.append(media_item)

        last_client_id = ""
        for item in items:
            last_client_id = self._generate_client_id()
            msg = WeixinMessage(
                from_user_id="", to_user_id=to_user_id, client_id=last_client_id,
                message_type=MessageType.BOT, message_state=MessageState.FINISH,
                item_list=[item], context_token=context_token,
            )
            request = SendMessageRequest(message=msg)
            self._api_client.send_message(request)
        return last_client_id

    def send_typing(self, ilink_user_id: str, typing_ticket: str, is_typing: bool = True):
        request = SendTypingRequest(
            ilink_user_id=ilink_user_id, typing_ticket=typing_ticket,
            status=TypingStatus.TYPING if is_typing else TypingStatus.CANCEL,
        )
        self._api_client.send_typing(request)
